library/Fission/App/Domain/Initializer/Class.hs

```diff
@@ -0,0 +1,11 @@
+module Fission.App.Domain.Initializer.Class (Initializer (..)) where
+
+import           Fission.Prelude
+import           Fission.URL.DomainName.Types (DomainName)
+
+class Monad m => Initializer m where
+  -- | The default domain that's managed by Fission
```

Comment:
I'm confused how this goes about getting the default value such that it can be used 
`defaultDomainName <- AppDomain.initial`

Reply:
No problem! This is a fairly common pattern, so good to go over:

```haskell
initial :: m DomainName
```

This says that we gen get a domain name inside some monadic context. This may or may not be a constant, it depends on `m`.

```haskell
instance App.DomainName.Initializer TestContext where
  initial = return (DomainName "fake.com") -- i.e. Hardcoded constant still possible

example = do
  domain <- App.DomainName.initial
  return (domain == DomanName "fake.com") -- True
```

The `Fission` monad contains environment variables it reads on startup, which is why this

1. isn't a hardcoded constant
2. needs to be in a monad (the "effect" is reading from the app's config)

```haskell
-- | The top level 'Fission' application 'RIO' configuration
data Config = Config
  {  baseAppDomainName :: !URL.DomainName
  -- ...plus a bunch more fields
  }

-- | The top-level app type
newtype Fission a = Fission { unwrapFission :: RIO Config a }
                                                -- ^^^^^^
                                                --  THIS

instance App.Domain.Initializer Fission where
  initial = asks baseAppDomainName -- i.e. extract the value from Config
```

---

To expand your mind on what you _can_ do with this:

Either is a monad for things that may fail.

```haskell
instance App.Domain.Initializer (Either MyError) where
  initial = Right (DomainName "example.com")
```

Lists are monads (for when you have multiple possible values)

```haskell
instance App.Domain.Initializer [a] where
  initial = [ DomainName "example.com"
            , DomainName "omgubuntu.co.uk"
            , DomainName "cbc.ca"
            ]
```

Random values:

```haskell
instance App.Domain.Initializer Random where
  initial = do
    txt <- randomText
    return (DomainName (txt <> ".com"))
```

Reply:
ahh okay this makes sense. I didn't see the Initializer instances in `Fission.Types` so was hopelessly confused how it was finding them :laughing: